dkg/pedersen/handler.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		return xerrors.Errorf("failed to create new DKG: %v", err)
+	}
+	h.dkg = d
+
+	// 2. Send my Deals to the other nodes
+	deals, err := d.Deals()
+	if err != nil {
+		return xerrors.Errorf("failed to compute the deals: %v", err)
+	}
+
+	// use a waitgroup to send all the deals asynchronously and wait
+	var wg sync.WaitGroup
+	wg.Add(len(deals))
+
+	for i, deal := range deals {
+		dealMsg := &Deal{
+			Index: deal.Index,
+			EncryptedDeal: &Deal_EncryptedDeal{
+				Dhkey:     deal.Deal.DHKey,
+				Signature: deal.Deal.Signature,
+				Nonce:     deal.Deal.Nonce,
+				Cipher:    deal.Deal.Cipher,
+			},
+			Signature: deal.Signature,
+		}
+
+		errs := out.Send(dealMsg, addrs[i])
+		go func(errs <-chan error) {
+			err, more := <-errs
+			if more {
+				fabric.Logger.Warn().Msgf("got an error while sending deal: %v", err)
+			}
+			wg.Done()
+		}(errs)
+	}
+
+	wg.Wait()
+
+	fabric.Logger.Trace().Msgf("%s sent all its deals", h.me)
+
+	receivedResps := make([]*pedersen.Response, 0)
+
+	numReceivedDeals := 0
+
+	// Process the deals we received before the start message
+	for _, deal := range receivedDeals {
+		err = h.handleDeal(deal, from, addrs, out)
+		if err != nil {
+			fabric.Logger.Warn().Msgf("%s failed to handle received deal "+
+				"from %s: %v", h.me, from, err)
+		}
+		numReceivedDeals++
+	}
+
+	// If there are N nodes, then N nodes first send (N-1) Deals. Then each node
+	// send a response to every other nodes. So the number of responses a node
+	// get is (N-1) * (N-1), where (N-1) should equal len(deals).
+	for numReceivedDeals < len(deals) {
+		from, msg, err := in.Recv(context.Background())
+		if err != nil {
+			return xerrors.Errorf("failed to receive after sending deals: %v", err)
+		}
+
+		switch msg := msg.(type) {
+
+		case *Deal:
+			// 4. Process the Deal and Send the response to all the other nodes
+			err = h.handleDeal(msg, from, addrs, out)
+			if err != nil {
+				fabric.Logger.Warn().Msgf("%s failed to handle received deal "+
+					"from %s: %v", h.me, from, err)
+			}
+			numReceivedDeals++
+
+		case *Response:
+			// 5. Processing responses
+			fabric.Logger.Trace().Msgf("%s received response from %s", h.me, from)
+			response := &pedersen.Response{
+				Index: msg.Index,
+				Response: &vss.Response{
+					SessionID: msg.Response.SessionID,
+					Index:     msg.Response.Index,
+					Status:    msg.Response.Status,
+					Signature: msg.Response.Signature,
+				},
+			}
+			receivedResps = append(receivedResps, response)
+
+		default:
+			return xerrors.Errorf("undexpected message: %T", msg)
+		}
+	}
+
+	for _, response := range receivedResps {
+		_, err = h.dkg.ProcessResponse(response)
+		if err != nil {
+			fabric.Logger.Warn().Msgf("%s failed to process response "+
+				"from '%s': %v", h.me, from, err)
+		}
+	}
+
+	for !h.dkg.Certified() {
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		from, msg, err := in.Recv(ctx)
```

Comment:
In that case, you expect the context to have some timeout if one of the N-1 node cannot answer for instance. Also the cancel is only triggered at the end of the function so it's not really useful here. The context is there to stop listening sooner if something happens. Here it seems to me a timeout is the only event that could happen.